File: app/src/main/java/com/radardeal/app/ui/detail/ListingDetailScreen.kt
package com.radardeal.app.ui.detail

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.rounded.ArrowBack
import androidx.compose.material.icons.rounded.Favorite
import androidx.compose.material.icons.rounded.FavoriteBorder
import androidx.compose.material.icons.rounded.OpenInNew
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.radardeal.app.core.Formatters
import com.radardeal.app.domain.model.DealRating
import com.radardeal.app.ui.components.BadgePill
import com.radardeal.app.ui.components.ListingImage
import com.radardeal.app.ui.components.PrimaryButton
import com.radardeal.app.ui.components.RdCard
import com.radardeal.app.ui.components.SectionHeader
import com.radardeal.app.ui.radar.openExternally
import com.radardeal.app.ui.radarViewModel
import com.radardeal.app.ui.theme.LocalSpacing
import com.radardeal.app.ui.theme.RadarColors
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

private val PhotoShape = RoundedCornerShape(24.dp)

/** Full view of one listing: photo, price, deal context and the recorded price history. */
@Composable
fun ListingDetailScreen(
    watchId: Long,
    itemId: String,
    onBack: () -> Unit,
) {
    val spacing = LocalSpacing.current
    val context = LocalContext.current
    val viewModel = radarViewModel(key = "detail-$watchId-$itemId") { graph ->
        ListingDetailViewModel(
            watchId = watchId,
            itemId = itemId,
            listingRepository = graph.listingRepository,
            watchRepository = graph.watchRepository,
        )
    }
    val state by viewModel.state.collectAsStateWithLifecycle()
    val listing = state.listing

    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(RadarColors.Background),
    ) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .statusBarsPadding()
                .padding(horizontal = spacing.sm, vertical = spacing.sm),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            IconButton(onClick = onBack) {
                Icon(
                    Icons.Rounded.ArrowBack,
                    contentDescription = "Retour",
                    tint = RadarColors.TextPrimary,
                )
            }
            Spacer(Modifier.weight(1f))
            if (listing != null) {
                IconButton(onClick = viewModel::toggleFavorite) {
                    Icon(
                        imageVector = if (listing.isFavorite) {
                            Icons.Rounded.Favorite
                        } else {
                            Icons.Rounded.FavoriteBorder
                        },
                        contentDescription = "Favori",
                        tint = if (listing.isFavorite) RadarColors.Danger else RadarColors.TextPrimary,
                    )
                }
            }
        }

        if (listing == null) {
            Box(
                modifier = Modifier.fillMaxSize(),
                contentAlignment = Alignment.Center,
            ) {
                Text(
                    text = if (state.isLoading) "Chargement…" else "Cette annonce n'est plus enregistrée.",
                    style = MaterialTheme.typography.bodyLarge,
                    color = RadarColors.TextSecondary,
                )
            }
            return@Column
        }

        Column(
            modifier = Modifier
                .weight(1f)
                .verticalScroll(rememberScrollState())
                .padding(horizontal = spacing.screen),
            verticalArrangement = Arrangement.spacedBy(spacing.lg),
        ) {
            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .aspectRatio(1f)
                    .clip(PhotoShape)
                    .background(RadarColors.SurfaceElevated),
            ) {
                ListingImage(url = listing.imageUrl, modifier = Modifier.fillMaxSize())
            }

            Row(horizontalArrangement = Arrangement.spacedBy(spacing.sm)) {
                if (listing.isNew) BadgePill(text = "NOUVEAU")
                if (listing.hasPriceDrop) {
                    BadgePill(
                        text = "PRIX ↓",
                        contentColor = RadarColors.PriceDrop,
                        background = RadarColors.PriceDropSoft,
                    )
                }
                if (state.rating != DealRating.NONE) {
                    BadgePill(text = state.rating.label)
                }
            }

            Text(
                text = listing.displayTitle,
                style = MaterialTheme.typography.headlineLarge,
                color = RadarColors.TextPrimary,
            )

            listing.subtitle().takeIf { it.isNotBlank() }?.let {
                Text(
                    text = it,
                    style = MaterialTheme.typography.bodyLarge,
                    color = RadarColors.TextSecondary,
                )
            }

            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(spacing.md),
            ) {
                Text(
                    text = Formatters.price(listing.price, listing.currency),
                    style = MaterialTheme.typography.displayMedium,
                    color = RadarColors.TextPrimary,
                )
                listing.previousPrice?.takeIf { listing.hasPriceDrop }?.let { previous ->
                    Column {
                        Text(
                            text = Formatters.price(previous, listing.currency),
                            style = MaterialTheme.typography.bodyMedium,
                            color = RadarColors.TextTertiary,
                            textDecoration = androidx.compose.ui.text.style.TextDecoration.LineThrough,
                        )
                        listing.priceDropRatio?.let { ratio ->
                            Text(
                                text = "-${(ratio * 100).toInt()} %",
                                style = MaterialTheme.typography.titleMedium,
                                color = RadarColors.PriceDrop,
                            )
                        }
                    }
                }
            }

            state.discountLabel?.let { label ->
                RdCard(
                    modifier = Modifier.fillMaxWidth(),
                    color = RadarColors.AccentSoft,
                    border = null,
                ) {
                    Column(
                        modifier = Modifier.padding(spacing.lg),
                        verticalArrangement = Arrangement.spacedBy(spacing.xs),
                    ) {
                        Text(
                            text = label,
                            style = MaterialTheme.typography.titleLarge,
                            color = RadarColors.Accent,
                        )
                        Text(
                            text = "Comparé aux ${state.comparableCount} annonces détectées par " +
                                "la veille « ${state.watchName} ». Ce n'est pas un prix de marché : " +
                                "c'est la médiane de ce que RadarDeal a vu passer.",
                            style = MaterialTheme.typography.bodyMedium,
                            color = RadarColors.TextSecondary,
                        )
                    }
                }
            }

            SectionHeader(title = "Historique")

            RdCard(modifier = Modifier.fillMaxWidth()) {
                Column(
                    modifier = Modifier.padding(spacing.lg),
                    verticalArrangement = Arrangement.spacedBy(spacing.sm),
                ) {
                    DetailRow("Veille", state.watchName)
                    DetailRow("Première détection", absoluteTime(listing.firstSeenAt))
                    DetailRow("Vue pour la dernière fois", Formatters.relativeTime(listing.lastSeenAt))

                    if (state.priceHistory.size > 1) {
                        Spacer(Modifier.height(spacing.xs))
                        Text(
                            text = "Prix relevés",
                            style = MaterialTheme.typography.titleMedium,
                            color = RadarColors.TextPrimary,
                        )
                        state.priceHistory.asReversed().take(8).forEach { point ->
                            DetailRow(
                                label = absoluteTime(point.recordedAt),
                                value = Formatters.price(point.price, listing.currency),
                            )
                        }
                    }
                }
            }

            Spacer(Modifier.height(spacing.sm))
        }

        Column(
            modifier = Modifier
                .padding(horizontal = spacing.screen)
                .padding(top = spacing.md, bottom = spacing.lg)
                .navigationBarsPadding(),
        ) {
            PrimaryButton(
                text = "Voir sur Vinted",
                onClick = { openExternally(context, listing.itemUrl) },
                icon = Icons.Rounded.OpenInNew,
            )
        }
    }
}

@Composable
private fun DetailRow(label: String, value: String) {
    Row(
        modifier = Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.SpaceBetween,
    ) {
        Text(
            text = label,
            style = MaterialTheme.typography.bodyMedium,
            color = RadarColors.TextSecondary,
        )
        Text(
            text = value,
            style = MaterialTheme.typography.titleMedium,
            color = RadarColors.TextPrimary,
        )
    }
}

private fun absoluteTime(timestamp: Long): String = runCatching {
    SimpleDateFormat("d MMM yyyy · HH:mm", Locale.FRANCE).format(Date(timestamp))
}.getOrDefault("—")

File: ios/App/RadarDeal/Screens/ListingDetailScreen.swift
import SwiftUI
import RadarDealCore

/// Full view of one listing: photo, price, deal context and the recorded price history.
struct ListingDetailScreen: View {
    let listing: Listing

    @Environment(AppGraph.self) private var graph

    @State private var current: Listing?
    @State private var assessment = DealEngine.Assessment.none
    @State private var history: [PricePoint] = []
    @State private var watchName = ""

    private var shown: Listing { current ?? listing }

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: RD.Space.lg) {
                ListingImage(url: shown.imageURL)
                    .aspectRatio(1, contentMode: .fill)
                    .frame(maxWidth: .infinity)
                    .clipped()
                    .clipShape(RoundedRectangle(cornerRadius: RD.Radius.card, style: .continuous))

                HStack(spacing: RD.Space.sm) {
                    if shown.isNew { BadgePill(text: "NOUVEAU") }
                    if shown.hasPriceDrop {
                        BadgePill(text: "PRIX ↓", foreground: RD.priceDrop,
                                  background: RD.priceDrop.opacity(0.14))
                    }
                    if assessment.rating != .none { BadgePill(text: assessment.rating.label) }
                }

                Text(shown.displayTitle).font(.rdDisplayMedium).foregroundStyle(RD.textPrimary)

                if !shown.subtitle().isEmpty {
                    Text(shown.subtitle()).font(.rdBody).foregroundStyle(RD.textSecondary)
                }

                HStack(alignment: .firstTextBaseline, spacing: RD.Space.md) {
                    Text(Formatters.price(shown.price, currency: shown.currency))
                        .font(.rdDisplayMedium)
                        .foregroundStyle(RD.textPrimary)

                    if shown.hasPriceDrop, let previous = shown.previousPrice {
                        VStack(alignment: .leading) {
                            Text(Formatters.price(previous, currency: shown.currency))
                                .font(.rdBodySmall)
                                .foregroundStyle(RD.textTertiary)
                                .strikethrough()
                            if let ratio = shown.priceDropRatio {
                                Text("-\(Int(ratio * 100)) %")
                                    .font(.rdTitleSmall)
                                    .foregroundStyle(RD.priceDrop)
                            }
                        }
                    }
                }

                if let discount = assessment.discount,
                   assessment.comparableCount >= DealEngine.minComparables {
                    RDCard(color: RD.accentSoft, bordered: false) {
                        VStack(alignment: .leading, spacing: RD.Space.xs) {
                            Text("\(Int(discount * 100)) % sous la médiane des annonces détectées")
                                .font(.rdTitle)
                                .foregroundStyle(RD.accent)
                            Text("Comparé aux \(assessment.comparableCount) annonces détectées "
                                 + "par la veille « \(watchName) ». Ce n'est pas un prix de "
                                 + "marché : c'est la médiane de ce que RadarDeal a vu passer.")
                                .font(.rdBodySmall)
                                .foregroundStyle(RD.textSecondary)
                                .fixedSize(horizontal: false, vertical: true)
                        }
                        .padding(RD.Space.lg)
                    }
                }

                SectionHeader(title: "Historique")

                RDCard {
                    VStack(alignment: .leading, spacing: RD.Space.sm) {
                        detailRow("Veille", watchName)
                        detailRow("Première détection", absolute(shown.firstSeenAt))
                        detailRow("Vue pour la dernière fois",
                                  Formatters.relativeTime(shown.lastSeenAt))

                        if history.count > 1 {
                            Text("Prix relevés").font(.rdTitleSmall)
                                .foregroundStyle(RD.textPrimary)
                                .padding(.top, RD.Space.xs)
                            ForEach(history.reversed().prefix(8), id: \.recordedAt) { point in
                                detailRow(
                                    absolute(point.recordedAt),
                                    Formatters.price(point.price, currency: shown.currency)
                                )
                            }
                        }
                    }
                    .padding(RD.Space.lg)
                }
            }
            .padding(.horizontal, RD.Space.screen)
            .padding(.bottom, RD.Space.xxl)
        }
        .background(RD.background)
        .scrollContentBackground(.hidden)
        .safeAreaInset(edge: .bottom) {
            if let url = shown.itemURL, let link = URL(string: url) {
                Link(destination: link) {
                    HStack(spacing: 8) {
                        Image(systemName: "arrow.up.right.square")
                        Text("Voir sur Vinted").font(.rdTitle)
                    }
                    .frame(maxWidth: .infinity, minHeight: 54)
                    .foregroundStyle(RD.onAccent)
                    .background(RD.accent, in: Capsule())
                }
                .padding(.horizontal, RD.Space.screen)
                .padding(.vertical, RD.Space.md)
                .background(RD.background)
            }
        }
        .toolbar {
            ToolbarItem(placement: .topBarTrailing) {
                Button {
                    graph.store.setFavorite(watchID: shown.watchID, itemID: shown.itemID,
                                            favorite: !shown.isFavorite)
                    reload()
                } label: {
                    Image(systemName: shown.isFavorite ? "heart.fill" : "heart")
                        .foregroundStyle(shown.isFavorite ? RD.danger : RD.textPrimary)
                }
            }
        }
        .task {
            // Opening a listing is what marks it as seen.
            graph.store.clearNewFlag(watchID: listing.watchID, itemID: listing.itemID)
            reload()
        }
    }

    private func detailRow(_ label: String, _ value: String) -> some View {
        HStack {
            Text(label).font(.rdBodySmall).foregroundStyle(RD.textSecondary)
            Spacer()
            Text(value).font(.rdTitleSmall).foregroundStyle(RD.textPrimary)
        }
    }

    private func reload() {
        current = graph.store.listings(watchID: listing.watchID)
            .first { $0.itemID == listing.itemID }
        watchName = graph.store.watch(id: listing.watchID)?.name ?? "Veille supprimée"
        history = graph.store.priceHistory(watchID: listing.watchID, itemID: listing.itemID)
        assessment = DealEngine.assess(
            price: shown.price, allPrices: graph.store.prices(watchID: listing.watchID)
        )
    }

    private func absolute(_ date: Date) -> String {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "fr_FR")
        formatter.dateFormat = "d MMM yyyy · HH:mm"
        return formatter.string(from: date)
    }
}
